src/Storages/IRemoteFileMetadata.h

```diff
@@ -0,0 +1,78 @@
+#pragma once
+#include <memory>
+#include <base/types.h>
+#include <functional>
+#include <unordered_map>
+#include <boost/core/noncopyable.hpp>
+
+namespace DB
+{
+class IRemoteFileMetadata
+{
+public:
+    IRemoteFileMetadata() = default;
+    IRemoteFileMetadata(const String & remote_path_,
+            size_t file_size_,
+            UInt64 last_modification_timestamp_):
+        remote_path(remote_path_)
+        ,file_size(file_size_)
+        ,last_modification_timestamp(last_modification_timestamp_)
+    {
+    }
+    virtual ~IRemoteFileMetadata();
+    virtual String getName() const = 0; //class name
+    // methods for basic information
+    inline size_t getFileSize() const { return file_size; }
+    inline String getRemotePath() const { return remote_path; }
+    inline UInt64 getLastModificationTimestamp() const { return last_modification_timestamp; }
+
+    // deserialize
+    virtual bool fromString(const String & buf) = 0;
+    // serialize
+    virtual String toString() const = 0;
+
+    // used for comparing two file metadatas are the same or not.
+    virtual String getVersion() const = 0;
+protected:
+    String remote_path;
+    size_t file_size = 0;
+    UInt64 last_modification_timestamp = 0;
+};
+
+using IRemoteFileMetadataPtr = std::shared_ptr<IRemoteFileMetadata>;
+
+/*
+ * How to register a subclass into the factory and use it ?
+ * 1) define your own subclass derive from IRemoteFileMetadata. Notice! the getName() must be the same
+ *    as your subclass name.
+ * 2) in a .cpp file, call REGISTTER_REMOTE_FILE_META_DATA_CLASS(subclass),
+ * 3) call RemoteFileMetadataFactory::instance().createClass(subclass_name) where you want to make a new object
```

Comment:
Probably like this
```suggestion
 * 3) call RemoteFileMetadataFactory::instance().registerClass(subclass_name) where you want to make a new object
```
though I would rather call it `registerMetadata`

Reply:
This method is used to create a new object , not register a class

Reply:
The point was that there was no method called `createClass` or I missed where it was (but I saw registerClass method nearby so thought just a comment was outdated).
Anyway as I see it is not important now, as this part is reimplemented.